be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "runtime/runtime_filter_cache.h"
+
+#include <chrono>
+
+#include "gutil/strings/substitute.h"
+#include "util/thread.h"
+
+namespace starrocks {
+
+using std::chrono::seconds;
+using std::chrono::milliseconds;
+using std::chrono::steady_clock;
+using std::chrono::duration_cast;
+class RfCacheValue {
+public:
+    RfCacheValue() { _extend_lifetime(); }
+    ~RfCacheValue() = default;
+    void put_if_absent(int filter_id, const JoinRuntimeFilterPtr& filter) {
+        _extend_lifetime();
+        if (!_filters.count(filter_id)) {
+            _filters[filter_id] = filter;
+        }
+    }
+
+    JoinRuntimeFilterPtr get(int filter_id) const {
+        _extend_lifetime();
+        auto it = _filters.find(filter_id);
+        if (it != _filters.end()) {
+            return it->second;
+        } else {
+            return nullptr;
+        }
+    }
+
+    bool is_expired() const {
+        auto now = duration_cast<milliseconds>(steady_clock::now().time_since_epoch()).count();
+        return now > _deadline;
+    }
+
+private:
+    void _extend_lifetime() const {
+        _deadline = duration_cast<milliseconds>(steady_clock::now().time_since_epoch() + EXPIRE_SECONDS).count();
```

Comment:
Out of curiosity, Does this way to handle lifetime realiable? Is EXPIRE_SECONDS long enough?

Reply:
stead_clock is stable,  enough

Reply:
No, I mean, is `EXPIRE_SECONDS` long enough? Is it possible thar lifetime epxires, while query is still executing?

Reply:
it depends,  this is just the wanted behevior,  rf cache just pre-cache rf that its target instances have not arrived or prepared, A query's  fragment instances  should not spend more 60s to deliver and prepare. after OperatorFactory install the RFs from cache, the RFs in the cache is needless, it can cleaned.